### ISSUE TITLE
Scrapers: What information can we get?

### ISSUE THREAD

**Issue by dvenprasad** [AlexsLemonade/resources-portal#2]
### Context

Users should be able to list resources by providing an URL which the portal can scrape the basic information about the resource.

### Problem or idea

Repos to target for MVP:
[AddGene](https://www.addgene.org/) - Plasmids
[GEO](https://www.ncbi.nlm.nih.gov/geo/) - Datasets
[SRA](https://www.ncbi.nlm.nih.gov/sra) - Datasets
[ATCC](https://www.atcc.org/) - Cell Lines
[Jackson Lab](https://www.jax.org/)- Mice
[protocols.io](https://www.protocols.io/) - Protocols
[dbGaP](https://www.ncbi.nlm.nih.gov/gap/)- Datasets
[Zfin](http://zfin.org/)- Zebrafish

It will be useful to know what fields are getting scraped and have some sort of a dummy data so we can gain a better understanding of how to present it. 

### Solution or next step

Tagging @arielsvn @davidsmejia @kurtwheeler for exploration

**Comment by dvenprasad** [comment]
Also, what stats can we scrape?

**Comment by cgreene** [comment]
For each one:

* A sample response
* How hard it is to derive that response?
* Does the API (if any) have documentation

Next steps: @dvenprasad will look into the specific endpoint listing relevant resources for the mice, zebrafish, and cell lines

GEO -> Series
SRA -> SRP

**Comment by dvenprasad** [comment]
**Cell lines- ATTC**
https://www.atcc.org/products/all/CRL-2551.aspx
https://www.atcc.org/products/all/CRL-1582.aspx
https://www.atcc.org/products/all/CCL-131.aspx

**Zebrafish Zfin(zric)** 
http://zfin.org/ZDB-GENO-080606-164
http://zfin.org/ZDB-GENO-070209-167


**Mouse jax lab** 
https://www.jax.org/strain/000686
https://www.jax.org/strain/002216
https://www.jax.org/strain/002288

**Comment by arielsvn** [comment]
Some notes after a discussion with @dvenprasad 

The most important fields for each resource type are

- [Plasmid](https://miro.com/app/board/o9J_kwI2PHs=/?moveToWidget=3074457346997702393)
- [Datasets](https://miro.com/app/board/o9J_kwI2PHs=/?moveToWidget=3074457346998016982)
- [Scientific Source Code](https://miro.com/app/board/o9J_kwI2PHs=/?moveToWidget=3074457346998104708)
- [Cell Lines](https://miro.com/app/board/o9J_kwI2PHs=/?moveToWidget=3074457346998017229)
- [Protocols](https://miro.com/app/board/o9J_kwI2PHs=/?moveToWidget=3074457346998227589)
- [Bio Samples](https://miro.com/app/board/o9J_kwI2PHs=/?moveToWidget=3074457346998247828)
- [Model Organisms](https://miro.com/app/board/o9J_kwI2PHs=/?moveToWidget=3074457346997727657)

[SciCrunch.org](https://scicrunch.org/resources/about/guidelines#celllines) already scrapes some of this information. The source code is in https://github.com/SciCrunch/SciCrunch-Portal - however it's in PHP.

**Comment by arielsvn** [comment]
I looked into each one of these resources, for most I couldn't find a documented API. So if we want to use them we'll have to scrape the data directly from the website. I didn't add sample responses, for the ones that we need to scrape we can grab any information that's visible on the webpage. And because it's scrapping they will be hard to maintain if we decide to include them.

@dvenprasad @kurtwheeler @davidsmejia Let me know if you have any questions or ideas.

---

### [GEO](https://www.ncbi.nlm.nih.gov/geo/) and [SRA](https://www.ncbi.nlm.nih.gov/sra) - Datasets

These two we can scrape, they have an API that we already use in refine.bio

### [dbGaP](https://www.ncbi.nlm.nih.gov/gap/) - Datasets

I checked one of the [public datasets](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000291.v2.p1) and all metadata is available in an [FTP server](ftp://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs000291/phs000291.v2.p1/) in XML files. I couldn't find any tools to read it, but I think its something we can solve.

There's a tool called prefetch that can be used to download data with ASPERA. Instructions [here](https://www.ncbi.nlm.nih.gov/books/NBK36439/#Download.download_using_prefetch_command). We'll need a PI account to download protected datasets.

### [AddGene](https://www.addgene.org/) - Plasmids

If we want this information we'll have to scrape it. All the plasmids are listed [here](https://www.addgene.org/search/catalog/plasmids/). Not all plasmids have the same fields so this could become a challenge.

With no API I think importing from addgene could be hard to maintain, I would recommend linking out.

### [ATCC](https://www.atcc.org/) - Cell Lines

Couldn't find an API for this either. All Human cell lines listed [here](https://www.atcc.org/search?title=ATCC%20Human%20Cell%20Lines%20-%20Alphanumeric#q=%40productline%3DC021&sort=relevancy).

So we'll have to scrape this information from the web page. Looks like all of the main fields are there.

Also would recommend linking out for these resources, given that they don't have an API.

### [Jackson Lab](https://www.jax.org/)- Mice

Looks like they have an internal [API](https://www.jax.org/mouse-search/SearchDataSheet), I couldn't find any documentation for it though.

Their [search page](https://www.jax.org/mouse-search) uses it, an example request sends `POST` to `https://www.jax.org/mouse-search/SearchDataSheet`

```json
{
  "searchRequest": {
    "Criteria": [
      {
        "SearchCriteriaType": 2,
        "SearchCriteriaTypeDisplay": "Featured Searches",
        "Value": "Most Popular Strains",
        "afid": ""
      }
    ],
    "resultsTabIndex": 0,
    "orderBy": "sortbypopularity",
    "Page": 2,
    "itemsPerPage": "50"
  }
}
```

The results only return the title and a marketing description for each item, eg: `"MarketingDescription: "Mice homozygous for the lymphoproliferation spontaneous mutation (<i>Fas<sup>lpr</sup></i>) show systemic autoimmunity, massive lymphadenopathy associated with proliferation of aberrant T cells, arthritis, and immune complex glomerulonephrosis.  These mice serve as a model for systemic lupus erythematosus-like autoimmune syndromes."`

The [deailed pages](https://www.jax.org/strain/003724) don't use any API endpoints.

In short, even though they have an undocumented API I think if we wanted their data we would be better off scraping their website.

### [protocols.io](https://www.protocols.io/) - Protocols

This one is well explained in the original site, and we can link out to it.

### [Zfin](http://zfin.org/) - Zebrafish

They have an internal API endpoint for the [autocomplete](http://zfin.org/action/quicksearch/autocomplete?q=hda) but I couldn't find any for the search.

Looks like the best option would be scrapping from the [search page](http://zfin.org/search?q=hdac&fq=category%3A%22Fish%22&category=Fish). I would also recommend linking out for this resource type

**Comment by cgreene** [comment]
Since it looks like things will be variable, I wonder if the flow is something like:

* A form for users to add these external resources that asks for the ID
* User inputs an ID. We attempt to query the source database (for v1.1 it looks like that'd be SRA, GEO, protocols.io APIs)
* User goes to something like the standard "create resources" form where we use the results of our query to pre-fill in fields. If the query failed to return / failed to parse, the field is blank.

Is something like this flexible enough to save the user time filling in fields while being robust if the endpoints end up being brittle?

**Comment by arielsvn** [comment]
@dvenprasad is this information enough? Or do you think we should explore additional resources?

**Comment by arielsvn** [comment]
protocols.io does have an API https://apidoc.protocols.io/
